setup.ps1

```diff
@@ -0,0 +1,229 @@
+    Write-Log "No input provided for installation." -Level 'ERROR'
+    Exit-Script -exitCode 1
+  }
+
+  Write-Log "Installing $InputPath..."
+  $installCmd = if (Test-Path $InputPath) { $VenvPip + @('install', '-r', $InputPath) } else { $VenvPip + @('install', $InputPath) }
+  $output = & $installCmd[0] $installCmd[1..$installCmd.Length] 2>&1
+  $output | Out-File $LogFilePath -Append
+  if ($LASTEXITCODE -ne 0) {
+    Write-Log "Conflict detected. Exiting now..." -Level 'ERROR'
+    Exit-Script -exitCode 1
+  }
+}
+
+# Function to get the installed Python version tag for wheel compatibility
+function Get-PythonVersionTag {
+  $pythonVersion = & python -c "import sys; print(f'cp{sys.version_info.major}{sys.version_info.minor}')"
+  $architecture = & python -c "import platform; print('win_amd64' if platform.machine().endswith('64') else 'win32')"
+  return "$pythonVersion-$architecture"
+}
+
+# Exit in test environment
+if ($MyInvocation.InvocationName -ne $MyInvocation.MyCommand.Name) {
+  exit
+}
+
+# Check for admin privileges and elevate if necessary
+if (-NOT ([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]::Administrator)) {
+  Write-Log "Requesting administrative privileges..." -Level 'ERROR'
+  Start-Process PowerShell -ArgumentList "-File `"$PSCommandPath`"" -Verb RunAs
+  Exit-Script -exitCode 1 -isSubShell $false
+}
+
+# Log file setup
+"Admin rights confirmed" | Out-File $LogFilePath -Append
+"Starting the script operations..." | Out-File $LogFilePath -Append
+
+# Navigate to the project directory
+Set-Location -Path $ProjectDir
+"Current directory: $(Get-Location)" | Out-File $LogFilePath -Append
+
+# Define the path to the Python executable in the virtual environment
+$VenvPython = Join-Path $ProjectDir "$VenvName\Scripts\python.exe"
+
+# Check if the virtual environment exists, if not, create it
+if (-Not (Test-Path $VenvPython)) {
+  Write-Log "Virtual environment not found. Creating virtual environment..." -IsError $false
+  python -m venv "$VenvName"
+  if (-Not (Test-Path $VenvPython)) {
+    Write-Log "Failed to create virtual environment." -Level 'ERROR'
+    Exit-Script -exitCode 1
+  }
+  Write-Log "Virtual environment created successfully." -IsError $false
+}
+
+# Define the pip command using the Python executable
+$VenvPip = @($VenvPython, '-m', 'pip')
+
+# Activate the virtual environment
+$OldVirtualPath = $env:PATH
+$env:PATH = "$ProjectDir\$VenvName\Scripts;$env:PATH"
+
+# Ensure setuptools is installed using the virtual environment's pip
+Write-Log "Ensuring setuptools is installed..."
+& $VenvPip[0] $VenvPip[1..$VenvPip.Length] 'install', '-v', 'setuptools' | Out-File $LogFilePath -Append 2>&1
+if ($LASTEXITCODE -ne 0) {
+  Write-Log "Failed to install setuptools." -Level 'ERROR'
+  Exit-Script -exitCode 1
+}
+
+# Pull latest updates
+Write-Log "Pulling latest updates..."
+& git pull | Out-File $LogFilePath -Append 2>&1
```

Comment:
I think people will only run it twice to update and run it once for initial installation. If it fails because the git state is dirty, it isn't a problem, because then you don't want the update. Better would be to check if the git state is dirty and only trigger the update if it is not the case. If we remove the git pull, people will not get the option to update. If we add a question, it makes the script more complex from UX perspective than necessary.